File: packages/chat-sync/src/chat-history.test.ts
import { describe, expect, test } from 'bun:test';
import type { FlueConversationMessage } from '@flue/sdk';

import {
  changedMessagesForConvexSync,
  mergeConvexMessages,
  messagesForConvexSync,
} from './index';

function message(
  id: string,
  role: 'user' | 'assistant',
  text: string,
  submissionId?: string,
): FlueConversationMessage & { role: 'user' | 'assistant' } {
  const built: FlueConversationMessage & { role: 'user' | 'assistant' } = {
    id,
    role,
    purpose: role,
    display: 'visible',
    parts: [{ type: 'text', text, state: 'done' }],
  };
  if (submissionId) built.submissionId = submissionId;
  return built;
}

describe('chat history', () => {
  test('keeps distinct user and assistant turns that share a submission', () => {
    const user = message('user:1', 'user', 'Hello', 'submission:1');
    const assistant = message(
      'assistant:1',
      'assistant',
      'Hi!',
      'submission:1',
    );

    expect(mergeConvexMessages([], [user, assistant])).toEqual([
      user,
      assistant,
    ]);
  });

  test('waits for admission before giving a local user turn a durable id', () => {
    const pending = message('local:1', 'user', 'Hello');
    const admitted = message('local:1', 'user', 'Hello', 'submission:1');

    expect(messagesForConvexSync([pending])).toEqual([]);
    expect(messagesForConvexSync([admitted])).toEqual([
      { ...admitted, id: 'submission:submission:1' },
    ]);
  });

  test('lets each live streaming envelope replace its stale durable copy', () => {
    const durable = message('assistant:1', 'assistant', 'Hel');
    const first = message('assistant:1', 'assistant', 'Hello');
    const second = message('assistant:1', 'assistant', 'Hello from Bee');
    const row = {
      id: durable.id,
      contentJson: JSON.stringify(durable),
      createdAt: 1,
    };

    expect(mergeConvexMessages([row], [first])).toEqual([first]);
    expect(mergeConvexMessages([row], [second])).toEqual([second]);
  });

  test('drops tombstoned rows and their live Flue twins after a retry', () => {
    const oldUser = message('user:1', 'user', 'What is my wallet?', 'sub-1');
    const oldAssistant = message('assistant:1', 'assistant', 'It failed.');
    const newUser = message('user:2', 'user', 'What is my wallet?', 'sub-2');
    const newAssistant = message('assistant:2', 'assistant', 'Here it is.');
    const rows = [
      {
        id: 'submission:sub-1',
        contentJson: JSON.stringify({ ...oldUser, id: 'submission:sub-1' }),
        createdAt: 1,
        hidden: true,
      },
      {
        id: oldAssistant.id,
        contentJson: JSON.stringify(oldAssistant),
        createdAt: 2,
        hidden: true,
      },
      {
        id: 'submission:sub-2',
        contentJson: JSON.stringify({ ...newUser, id: 'submission:sub-2' }),
        createdAt: 3,
      },
    ];

    // The live Flue transcript still contains the retried turn; hidden rows
    // must suppress both the durable copy and the local echo.
    expect(
      mergeConvexMessages(rows, [
        oldUser,
        oldAssistant,
        newUser,
        newAssistant,
      ]),
    ).toEqual([newUser, newAssistant]);
  });

  test('returns only changed live envelopes after the initial snapshot', () => {
    const user = message('user:1', 'user', 'Hello', 'submission:1');
    const partial = message('assistant:1', 'assistant', 'Hel');
    const initial = changedMessagesForConvexSync([user, partial], new Map());
    const known = new Map(
      initial.map((entry) => [entry.id, entry.contentJson]),
    );
    const next = message('assistant:1', 'assistant', 'Hello');

    expect(changedMessagesForConvexSync([user, partial], known)).toEqual([]);
    expect(changedMessagesForConvexSync([user, next], known)).toEqual([
      expect.objectContaining({
        id: next.id,
        contentJson: JSON.stringify(next),
      }),
    ]);
  });
});

File: apps/android/core/contract/src/test/kotlin/com/beegreat/contract/ChatHistoryTest.kt
package com.beegreat.contract

import com.beegreat.flue.FlueMessage
import com.beegreat.flue.FluePart
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import org.junit.Assert.assertEquals
import org.junit.Test

class ChatHistoryTest {
  private fun user(id: String, text: String, submissionId: String? = null, timestamp: String? = null) =
    FlueMessage(
      id = id,
      role = "user",
      purpose = "user",
      submissionId = submissionId,
      parts = listOf(FluePart.Text(text, "done")),
      metadata = timestamp?.let { JsonObject(mapOf("timestamp" to JsonPrimitive(it))) },
    )

  private fun assistant(id: String, text: String, state: String = "done") =
    FlueMessage(id = id, role = "assistant", purpose = "assistant", parts = listOf(FluePart.Text(text, state)))

  private fun row(message: FlueMessage, createdAt: Long, hidden: Boolean = false) =
    StoredChatMessage(message.id, encodeMessage(message), createdAt, hidden)

  @Test
  fun `stored rows come first, live messages replace their stored copies by key`() {
    val storedUser = user("submission:s1", "hi", submissionId = "s1")
    val storedReply = assistant("a1", "hello")
    val live = listOf(user("local:1", "hi", submissionId = "s1"), assistant("a1", "hello there", "streaming"))
    val merged = mergeConvexMessages(listOf(row(storedUser, 1), row(storedReply, 2)), live)
    assertEquals(listOf("local:1", "a1"), merged.map { it.id })
    assertEquals("hello there", merged[1].text)
  }

  @Test
  fun `hidden rows tombstone both key forms and system messages never render`() {
    val retried = user("submission:s2", "again", submissionId = "s2")
    val system = FlueMessage(id = "sys", role = "system", purpose = "advisory", display = "diagnostic")
    val merged = mergeConvexMessages(listOf(row(retried, 5, hidden = true)), listOf(user("u2", "again", submissionId = "s2"), system))
    assertEquals(emptyList<String>(), merged.map { it.id })
  }

  @Test
  fun `metadata timestamps order messages, missing ones fall back to arrival`() {
    val late = assistant("late", "x").copy(metadata = JsonObject(mapOf("timestamp" to JsonPrimitive("2026-01-02T00:00:00Z"))))
    val early = user("early", "y", timestamp = "2026-01-01T00:00:00Z")
    val merged = mergeConvexMessages(listOf(row(late, 100)), listOf(early))
    assertEquals(listOf("early", "late"), merged.map { it.id })
  }

  @Test
  fun `changed detection keys admitted user turns by submission and skips local echoes`() {
    val messages = listOf(user("local:1", "typing"), user("local:2", "sent", submissionId = "s9"), assistant("a", "reply"))
    val changed = changedMessagesForConvexSync(messages, emptyMap())
    assertEquals(listOf("submission:s9", "a"), changed.map { it.id })
    val known = changed.associate { it.id to it.contentJson }
    assertEquals(emptyList<ChatMessageSyncEnvelope>(), changedMessagesForConvexSync(messages, known))
  }
}
